components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,211 @@
+        self.ranges_evcited.append(&mut other.ranges_evcited);
+        self.ranges_unreadable.append(&mut other.ranges_unreadable);
+
+        // merge safe_ts to the min of them if not 0
+        if self.safe_ts != 0 && other.safe_ts != 0 {
+            self.safe_ts = u64::min(self.safe_ts, other.safe_ts);
+        } else if other.safe_ts != 0 {
+            self.safe_ts = other.safe_ts;
+        }
+    }
+
+    pub(crate) fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        if set_readable {
+            assert!(self.ranges_unreadable.remove(range));
+        } else {
+            self.ranges_unreadable.insert(range.clone());
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RangeManager {
+    // the range reflects the range of data that is accessable.
+    ranges: BTreeMap<CacheRange, RangeMeta>,
+}
+
+impl RangeManager {
+    pub(crate) fn ranges(&self) -> &BTreeMap<CacheRange, RangeMeta> {
+        &self.ranges
+    }
+
+    pub(crate) fn new_range(&mut self, range: CacheRange) {
+        if let Some(sibling) = self.ranges.keys().find(|r| r.is_sibling(&range)).cloned() {
+            let left_sib = sibling < range;
+            let mut sib_meta = self.ranges.remove(&sibling).unwrap();
+            sib_meta.ranges_unreadable.insert(range.clone());
+            let mut range = CacheRange::merge(sibling, range);
+
+            // if sibling found above is the left sibling of the range, there could be a
+            // right sibling
+            if left_sib {
+                if let Some(right_sibling) =
+                    self.ranges.keys().find(|r| r.is_sibling(&range)).cloned()
+                {
+                    let right_sib_meta = self.ranges.remove(&right_sibling).unwrap();
+                    sib_meta.merge_meta(right_sib_meta);
+                    range = CacheRange::merge(range, right_sibling);
+                }
+            }
+            self.ranges.insert(range, sib_meta);
+        } else {
+            let mut range_meta = RangeMeta::default();
+            range_meta.ranges_unreadable.insert(range.clone());
+            self.ranges.insert(range, range_meta);
+        }
+    }
+
+    pub fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        let range_key = self
+            .ranges
+            .keys()
+            .find(|&r| r.contains(range))
+            .unwrap()
+            .clone();
+        let meta = self.ranges.get_mut(&range_key).unwrap();
+        meta.set_range_readable(range, set_readable);
+    }
+
+    pub fn set_safe_ts(&mut self, range: &CacheRange, safe_ts: u64) -> bool {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            if meta.safe_ts > safe_ts {
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
```

Comment:
Should we add contains method to indicate range is completely covered by a cached range?

Reply:
In fact, we should evict the range (or add the evict hint) if it just partially covers the input range.

Reply:
This is now used for assertion purpose where assert we are not create a range overlapping with existing ranges.